src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -177,5 +209,338 @@ private void ThrowIfDisposed()
+                topic = await _managementClient.GetTopicAsync(topicPath);
+                if (topic.EnablePartitioning)
+                {
+                    partitionCount = 16;
+                }
+            }
+            catch (MessagingEntityNotFoundException)
+            {
+                _logger.LogWarning($"ServiceBus subscription with path '{_entityPath}' was not found.");
+            }
+            catch (UnauthorizedAccessException) // When Manage claim is not used on Service Bus connection string
+            {
+                if (TimeToLogWarning())
+                {
+                    _logger.LogWarning($"Connection string does not have Manage claim for topic '{_entityPath}'. Failed to get subscription description to derive topic queue length metrics. " +
+                        $"Falling back to using first message enqueued time.");
+                }
+
+                isUnauthorized = true;
+            }
+            catch (Exception e)
+            {
+                _logger.LogWarning($"Error querying for Service Bus topic scale status: {e.Message}");
+            }
+
+            return CompileTriggerMetrics(message, activeMessageCount, deadLetterCount, partitionCount, isUnauthorized, _isListeningOnDeadLetterQueue);
+        }
+
+        public static ServiceBusTriggerMetrics CompileTriggerMetrics(Message message, long activeMessageCount, long deadLetterCount, int partitionCount,
+            bool isUnauthorized, bool isListeningOnDeadLetterQueue)
+        {
+            long totalNewEventCount = 0;
+            TimeSpan eventQueueTime = TimeSpan.Zero;
+
+            if (message != null)
+            {
+                eventQueueTime = DateTime.UtcNow.Subtract(message.SystemProperties.EnqueuedTimeUtc);
+                totalNewEventCount = 1;
+            }
+
+            if (!isUnauthorized && ((!isListeningOnDeadLetterQueue && activeMessageCount > 0) || (isListeningOnDeadLetterQueue && deadLetterCount > 0)))
+            {
+                totalNewEventCount = isListeningOnDeadLetterQueue ? deadLetterCount : activeMessageCount;
+            }
+
+            return new ServiceBusTriggerMetrics
+            {
+                Timestamp = DateTime.UtcNow,
+                TotalNewEventCount = totalNewEventCount,
+                PartitionCount = partitionCount,
+                EventQueueTime = eventQueueTime
+            };
+        }
+
+        ScaleStatus IScaleMonitor.GetScaleStatus(ScaleStatusContext context)
+        {
+            return GetScaleStatusCore(context.WorkerCount, context.Metrics?.Cast<ServiceBusTriggerMetrics>().ToArray());
+        }
+
+        public ScaleStatus GetScaleStatus(ScaleStatusContext<ServiceBusTriggerMetrics> context)
+        {
+            return GetScaleStatusCore(context.WorkerCount, context.Metrics?.ToArray());
+        }
+
+        private ScaleStatus GetScaleStatusCore(int workerCount, ServiceBusTriggerMetrics[] metrics)
```

Comment:
Is this code copy/paste from `ScaleController`?

Reply:
Yeah pretty much. I had to pick and choose which sections of the scale controller logic applied to service bus, but it all comes from ScaleController.